core/src/main/java/io/grpc/internal/RetryPolicy.java

```diff
@@ -0,0 +1,107 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Objects;
+import io.grpc.Status;
+import io.grpc.Status.Code;
+import java.util.Collections;
+import java.util.Set;
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * Retry policy data object.
+ */
+@Immutable
+final class RetryPolicy {
+  final int maxAttempts;
+  final long initialBackoffNanos;
+  final long maxBackoffNanos;
+  final double backoffMultiplier;
+  final Set<Code> retryableStatusCodes;
+
+  /** No retry. */
+  static final RetryPolicy DEFAULT =
+      new RetryPolicy(1, 0, 0, 1, Collections.<Status.Code>emptySet());
+
+  /**
+   * The caller is supposed to have validated the arguments and handled throwing exception or
+   * logging warnings already, so we avoid repeating args check here.
+   *
+   * @param retryableStatusCodes an immutable set.
+   */
+  RetryPolicy(
+      int maxAttempts,
+      long initialBackoffNanos,
+      long maxBackoffNanos,
+      double backoffMultiplier,
+      @Nonnull Set<Code> retryableStatusCodes) {
+    this.maxAttempts = maxAttempts;
+    this.initialBackoffNanos = initialBackoffNanos;
+    this.maxBackoffNanos = maxBackoffNanos;
+    this.backoffMultiplier = backoffMultiplier;
+    this.retryableStatusCodes = retryableStatusCodes;
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(
+        maxAttempts,
+        initialBackoffNanos,
+        maxBackoffNanos,
+        backoffMultiplier,
+        retryableStatusCodes);
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    if (!(other instanceof RetryPolicy)) {
+      return false;
+    }
+    RetryPolicy that = (RetryPolicy) other;
+    return Objects.equal(this.maxAttempts, that.maxAttempts)
+        && Objects.equal(this.initialBackoffNanos, that.initialBackoffNanos)
+        && Objects.equal(this.maxBackoffNanos, that.maxBackoffNanos)
+        && Double.compare(this.backoffMultiplier, that.backoffMultiplier) == 0
+        && Objects.equal(this.retryableStatusCodes, that.retryableStatusCodes);
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("maxAttempts", maxAttempts)
+        .add("initialBackoffNanos", initialBackoffNanos)
+        .add("maxBackoffNanos", maxBackoffNanos)
+        .add("backoffMultiplier", backoffMultiplier)
+        .add("retryableStatusCodes", retryableStatusCodes)
+        .toString();
+  }
+
+  /**
+   * Providers the most suitable retry policy for a call in the current situation.
```

Comment:
Without looking at DelayedRetryPolicyProvider, this description is hard to understand. Might be helpful to say when this will have to provide a retry policy, as that defines what the "current situation" will be